packages/runtime/runtime-utils/src/runtimeFactoryHelper.ts

```diff
@@ -0,0 +1,35 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+    IContainerContext,
+    IRuntime,
+    IRuntimeFactory,
+} from "@fluidframework/container-definitions";
+
+export abstract class RuntimeFactoryHelper implements IRuntimeFactory {
+    public get IRuntimeFactory() { return this; }
+
+    public async instantiateRuntime(context: IContainerContext, existing?: boolean): Promise<IRuntime> {
+        const fromExisting = existing === undefined
+            ? context.existing === true
+            : existing === true;
+        const runtime = await this.preInitialize(context, fromExisting);
+
+        if (fromExisting) {
+            await this.instantiateFromExisting(runtime);
+        } else {
+            await this.instantiateFirstTime(runtime);
+        }
+
+        await this.hasInitialized(runtime);
+        return runtime;
+    }
+
+    public abstract preInitialize(context: IContainerContext, existing: boolean): Promise<IRuntime>;
```

Comment:
IRuntime is not very useful interface to work with, as it looks upstream (to the loader), vs. methods like instantiateFromExisting wants to work with downstream interface (IContainerRuntime).

So we can make it return IRuntime & IContainerRuntime, which would more useful (and maybe drop IRuntime when calling methods on this class with resulting object)

We can also make it work with ContainerRuntime (concrete object), either directly or through making this class being generic (i.e. RuntimeFactoryHelper<T extends IContainerRuntime>). That might be good think if we want to get rid of IContainerRuntime (there is very little value in this interface, as code either works at this layer and knows it's ContainerRuntime, or code works with IContainerRuntimeBase - smaller version of this interface).
I'd probably defer doing this work until / if we make a choice to remove IContainerRuntime - using IContainerRuntime  for now should be sufficient.